BackendSelect: Use at::_ops API and per-operator headers

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #69840
* #70021
* #68693

Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

peterbell10 committed:
Codegen: Use per-operator headers in all registration templates

`RegisterSchema.cpp` only uses strings to register operator schemas,
so doesn't need to include any operator headers at all (except
indirectly through `torch/library.h`).

For `RegisterBackendSelect.cpp` I moved it over to the `at::_ops` API
and use the per-operator headers.

[ghstack-poisoned]

albanD commented on tools/codegen/gen.py:
```diff
@@ -643,11 +647,9 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
-    .typed<{dispatcher_sig.type()}>();
   {compute_dk}
-  return op.redispatch(_dk, {', '.join(a.expr for a in dispatcher_exprs)});
+  return at::_ops::{f.func.name.unambiguous_name()}::redispatch(
```
What's the justification for moving these calls from the static op to the at::_ops:: namespace?

peterbell10 replied:
I can back out the change if you want, but fwiw this is a binary size reduction on my machine. 2 MB before and 1.5 MB after using `at::_ops` which makes sense as you are removing all the magic statics code. 

AFAIK this is exactly what the `_ops` API is intended for as well. It makes the generated code simpler, removes a header dependency on the dispatcher internals, centralizes the `static` operator data into one place.

albanD replied:
Sounds good! Thanks for the clarification.

malfet replied:
I say it would probably be better to move it to separate PR, as it is not related to "Use per-operator headers" goal, does it?

peterbell10 replied:
I've split the changes across #70021 and updated the PR description to be explicit about using the `at::_ops` API.

albanD reviewed (commented):
Change looks ok. Not use about the move to use `at::_ops`.
These files are pretty sensitive in terms of binary size hopefully this should not change anything there?

peterbell10 committed:
Update on "BackendSelect: Use at::_ops API and per-operator headers"




[ghstack-poisoned]

peterbell10 committed:
Rebase to fix merge conflicts on "BackendSelect: Use at::_ops API and per-operator headers"




[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33160027).

peterbell10 committed:
Update on "BackendSelect: Use at::_ops API and per-operator headers"


Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33160027).

peterbell10 committed:
Update on "BackendSelect: Use at::_ops API and per-operator headers"


Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

[ghstack-poisoned]

peterbell10 committed:
Update on "BackendSelect: Use at::_ops API and per-operator headers"


Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33160027).

peterbell10 committed:
Update on "BackendSelect: Use at::_ops API and per-operator headers"


Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33160027).

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33160027).

peterbell10 committed:
Update on "BackendSelect: Use at::_ops API and per-operator headers"


Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33160027).

peterbell10 committed:
Rebase and fix merge conflicts on "BackendSelect: Use at::_ops API and per-operator headers"


Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33160027).

peterbell10 committed:
Update on "BackendSelect: Use at::_ops API and per-operator headers"


Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

[ghstack-poisoned]

peterbell10 committed:
Update on "BackendSelect: Use at::_ops API and per-operator headers"


Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33160027).

peterbell10 committed:
Update on "BackendSelect: Use at::_ops API and per-operator headers"


Differential Revision: [D33160027](https://our.internmc.facebook.com/intern/diff/D33160027)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33160027).
